fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/StandardIndexMaintainer.java

```diff
@@ -800,7 +800,9 @@ public CompletableFuture<Void> deleteWhere(Transaction tr, @Nonnull Tuple prefix
         // NOTE: Range.startsWith(), Subspace.range() and so on cover keys *strictly* within the range, but we sometimes
         // store data at the prefix key itself.
         final byte[] key = state.indexSubspace.pack(prefix);
-        tr.clear(key, ByteArrayUtil.strinc(key));
+        Range indexRange = new Range(key, ByteArrayUtil.strinc(key));
```

Comment:
I think the following also won't work:
1. `FDBRecordStore.deleteAllRecords`
2. `KeySpacePathImpl.deleteAllDataAsync`
3. `FDBRecordStore.rebuildAllIndexes`
4. `FDBRecordStore.clearIndexData`
5. `FDBRecordStore.removeFormerIndex`

I'm guessing this leans into what you were mentioning in the description that perhaps the abstractions are broken

I see that some other index maintainers also do range clears:
1. `TimeWindowLeaderboard`
2. `Multidimensional`
7. `Rank`
8. `PermutedMinMax`
Is there something to be concerned about here? I think, in theory, they should be able to support `version()` in some capacity, even if I'm not sure that it entirely makes sense. Or that their support is already covered by them calling `super.deleteWhere`.

Reply:
> I'm guessing this leans into what you were mentioning in the description that perhaps the abstractions are broken

Yeah, that's right. I've tried to rectify this a bit, by adding some new `clear` methods to `FDBRecordContext` that combine both clearing out the underlying data from FDB and cleaning up the version information stored in the record context. Then I switched over callers to this new method, including the methods you listed here. I've added additional tests to cover those, except for `rebuildAllIndexes`, which is a tough one given that the rebuilding of the index within the transaction should add back any of the deleted data.

> I see that some other index maintainers also do range clears. ... Is there something to be concerned about here?

None of the listed indexes currently support version mutations (and their `IndexValidator`s prevent them from being created with one), so those range clears shouldn't cause any additional problems. In most of those cases, the actual index format means it's impossible to order data correctly if there's a version column, with the possible exception of `PermutedMinMax`. That being said, switching over to the new method shouldn't be wrong and sets us up a bit better if we start to enable versions there in the future.